src/middleware.ts

```diff
@@ -1,106 +1,124 @@
             );
             for (const middlewareFunction of middlewareFunctions) {
-                await middlewareFunction(context, ...args);
+                await middlewareFunction(request, context, handlerResult);
             }
         }
 
-        if (error !== undefined) {
-            context.log.error(`An uncaught error occurred in the execution of the handler: ${error}`);
-            return { $failed: true, $error: error };
+        if (isErrorResult(handlerResult)) {
+            context.error(`An caught error occurred in the execution of the handler: ${stringify(handlerResult)}`);
+            return handlerResult.$error;
         }
-        return handlerResult;
+
+        if (handlerResult.$result === undefined) {
+            return new Error('Illegal-State: Result of the handler was empty.');
+        }
+
+        return handlerResult.$result;
     };
 
 export type Options = {
     errorResponseHandler?: (
         error: unknown,
-        context: Context,
+        context: InvocationContext,
     ) => {
         [key: string]: unknown;
     };
     disableErrorHandling?: boolean;
 };
 
-async function middlewareWrapper(
-    beforeExecution: (AzureFunction | false)[],
-    handler: (context: Context, ...args: unknown[]) => Promise<unknown> | void,
-    postExecution: (AzureFunction | false)[],
-    context: Context,
-    args: unknown[],
+async function middlewareWrapper<T extends FunctionHandler>(
+    beforeExecution: (BeforeExecutionFunction<T> | false)[],
+    handler: T,
+    postExecution: (PostExecutionFunction<T> | false)[],
+    request: Parameters<T>,
+    context: InvocationContext,
     opts?: Options,
 ) {
-    const result = await middlewareCore(beforeExecution, handler, postExecution)(context, ...args);
+    const result = await middlewareCore(beforeExecution, handler, postExecution)(request, context);
 
-    if (isErrorResult(result)) {
+    if (result instanceof Error) {
         if (opts?.disableErrorHandling) {
-            throw result.$error;
+            throw result;
         }
 
-        context.res = errorHandler(result.$error, context, opts);
-        return;
+        return errorHandler(result, context, opts);
     }
 
     return result;
 }
 
 export const middleware =
-    (
-        beforeExecution: (AzureFunction | false)[],
-        handler: AzureFunction,
-        postExecution: (AzureFunction | false)[],
+    <T extends FunctionHandler>(
+        beforeExecution: (BeforeExecutionFunction<T> | false)[],
+        handler: T,
+        postExecution: (PostExecutionFunction<T> | false)[],
         opts?: Options,
     ) =>
-    async (context: Context, ...args: unknown[]): Promise<unknown> => {
+    // eslint-disable-next-line @typescript-eslint/ban-ts-comment
+    // @ts-ignore - Sorry :( But FunctionResult<T = unknown> = T | Promise<T>; so it should be fine
+    async (request: Parameters<T>[0], context: InvocationContext): ReturnType<T> => {
```

Comment:
```suggestion
    async (request: Parameters<T>[0], context: InvocationContext): Promise<any> => {
```
Würde das nicht besser zu dem Typen aus Azure passen?